broker-daemon/orderbook/index.js

```diff
@@ -98,11 +100,17 @@ class Orderbook {
       watcher.removeListener('sync', onWatcherSync)
       watcher.removeListener('error', onWatcherError)
 
-      this.logger.info(`Market ${this.marketName} unavailable, retrying in ${RETRY_WATCHMARKET}ms`, { error })
-      // TODO: exponential backoff?
-      setTimeout(() => {
-        this.watchMarket()
-      }, RETRY_WATCHMARKET)
+      // Calculate the exponential backoff delay interval using the binary exponential backoff algorithm
+      // defined in https://en.wikipedia.org/wiki/Exponential_backoff
+      //
+      // Note: Repeated retransmission wait time is typically defined as a random variable between
+      // 0 and 2^c - 1, where c is the number of retries, however here we increase delay time using
+      // 2^c - 1 until MAX_RETRY_INTERVAL is reached.
+      retries++
+      const delay = Math.min((2 ** retries - 1) * 1000, MAX_RETRY_INTERVAL)
+
+      this.logger.info(`Market ${this.marketName} unavailable, retrying in ${delay}ms`, { error })
+      setTimeout(this.watchMarket, delay, retries)
```

Comment:
will `this.watchMarket` get called in the context of `this`? If not you may need to use a fat arrow function:

```
setTimeout(() => this.watchMarket(retries), delay)
```

Reply:
Yep, I just tried this branch locally and got:

```
sparkswapd_1  | error: Relayer watchMarket grpc failed {"code":14,"metadata":{"_internal_repr":{}},"details":"EOF","timestamp":"2019-01-08T19:34:21.914Z"}
sparkswapd_1  | info: Market BTC/LTC unavailable, retrying in 1000ms {"error":{"code":14,"metadata":{"_internal_repr":{}},"details":"EOF"},"timestamp":"2019-01-08T19:34:21.915Z"}
sparkswapd_1  | (node:100) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'debug' of undefined
sparkswapd_1  |     at Timeout.watchMarket [as _onTimeout] (/home/app/broker-daemon/orderbook/index.js:75:17)
sparkswapd_1  |     at ontimeout (timers.js:502:15)
sparkswapd_1  |     at tryOnTimeout (timers.js:323:5)
sparkswapd_1  |     at Timer.listOnTimeout (timers.js:290:5)
sparkswapd_1  | (node:100) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
```

which is due to `this.logger` not being defined in the given context.